src/dac.rs

```diff
@@ -0,0 +1,220 @@
+//! Configure the internal DAC on the stm32f3xx.
+//! Incomplete, but includes basic operation.
+//!
+//! You should have the approprite dac pin set up as an analog input, to prevent
+//! parasitic power consumption. For example:
+//! ```rust
+//! let _dac1_pin = gpioa.pa4.into_analog(&mut gpioa.moder, &mut gpioa.pupdr);
+//! let _dac2_pin = gpioa.pa5.into_analog(&mut gpioa.moder, &mut gpioa.pupdr);
+//! ```
+//!
+
+// todo: DAC trait that will go in a PR to embeddd hal.
+// [PR](https://github.com/rust-embedded/embedded-hal/pull/247)
+
+//! Digital-analog conversion traits
+//!
+//! A trait used to identify a digital-to-analog converter, and its
+//! most fundamental features.
+
+use core::fmt;
+
+use crate::{pac, rcc::APB1};
+
+pub trait SingleChannelDac {
+    /// Error type returned by DAC methods
+    type Error;
+    type Word;
+
+    /// Output a constant signal, given a bit word.
+    fn try_set_value(&mut self, value: Self::Word) -> Result<(), Self::Error>;
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum DacId {
+    One,
+    Two,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum DacBits {
+    EightR,
+    TwelveL,
+    TwelveR,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Trigger {
+    Tim6,
+    Tim3_8,
+    Tim7,
+    Tim15,
+    Tim2,
+    Tim4, // Not available on DAC 2.
+    Exti9,
+    Swtrig,
+}
+
+impl Trigger {
+    pub fn bits(&self) -> u8 {
+        match self {
+            Self::Tim6 => 0b000,
+            Self::Tim3_8 => 0b001,
+            Self::Tim7 => 0b010,
+            Self::Tim15 => 0b011,
+            Self::Tim2 => 0b100,
+            Self::Tim4 => 0b101,
+            Self::Exti9 => 0b110,
+            Self::Swtrig => 0b111,
+        }
+    }
+}
+
+pub struct Dac {
+    regs: pac::DAC,
+    id: DacId,
+    bits: DacBits,
+    vref: f32,
+}
+
+impl Dac {
+    /// Create a new DAC instances
+    pub fn new(regs: pac::DAC, id: DacId, bits: DacBits, vref: f32) -> Self {
+        Self {
+            regs,
+            id,
+            bits,
+            vref,
+        }
+    }
+
+    /// Enable the DAC.
+    pub fn enable(&mut self, apb1: &mut APB1) {
+        match self.id {
+            DacId::One => {
+                apb1.enr().modify(|_, w| w.dac1en().set_bit());
+                self.regs.cr.modify(|_, w| w.en1().set_bit());
```

Comment:
The PAC usually provides more readable methods then `set_bit`/`clear_bit`. Using these makes the code more readable and can even avoid bugs. In this case, the readable method would be `.enabled()`.